### ISSUE TITLE
Potential fatal error during Analytics account provisioning flow when setting up SK with Analytics 

### ISSUE THREAD

**Issue by mohitwp** [google/site-kit-wp#7124]
## Bug Description

I downloaded and installed the plugin from [latest pre-release build](https://storage.googleapis.com/site-kit-github-artifacts/builds/refs/heads/main/google-site-kit.zip) and after that I set up site kit with analytics. After creating analytics a/c when I click on 'Go to Dashboard' button then I'm getting Fatal Error. First it shows error page and when we refresh the page it shows Main dashboard. I verified this on 2 new sites. On both sites, I'm getting the same Fatal error. Also, verified with tester plugin after switching to main branch. Getting the same result.

## Steps to reproduce

1. Create new site.
2. Set up Site kit with analytics.
3. After creating analytics a/c click on "Go to Dashboard' button.
4. See Fatal error.

## Screenshots

![image](https://github.com/google/site-kit-wp/assets/94359491/901ce32d-fd34-4d38-bf52-eefcf8f5637c)

## Additional Context

- The error happens due to GTE related behavior that runs at the end of the Analytics account provisioning flow if the GTM scope has not been granted
- As of 1.101.0, the GTM scope is now required when setting up Analytics (#6806) but only when `gteSupport` is enabled
- When setting up Site Kit initially and choosing the option to "Connect Google Analytics as part of your setup", this activates the Analytics module and requires its scopes as part of the initial OAuth flow – before features are available to the site – so the GTM scope wouldn't be required
- The problematic GTE request was only guarded by the feature flag
- The problematic GTE request was made using the Google service directly instead of our internal data API which does not handle Google service exceptions, which is the source of this error
- In the event of this error, Analytics provisioning and module connection is successful and the user is not blocked 

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* No error should be raised during the Analytics account provisioning flow when `gteSupport` is enabled

## Implementation Brief

* Update GTE tag syncing in post-GA provisioning handler to use safer `get_data` method instead of calling API directly (this is sufficient to prevent the error)
	* This can use `get_data( 'google-tag-settings' )` passing through the measurement ID rather than doing the container lookup and then getting the tag settings
* Update required scopes for GA provisioning to include GTM scope if `gteSupport` is enabled and the scope isn't granted (this will ensure the GTE routine still runs post-provisioning)
    * In the event the scope is not granted as part of the flow, the container lookup request will still fail (not be made) due to the missing scope, but it won't raise the same error and the user will be prompted to grant it on the dashboard as usual

### Test Coverage

* Add test coverage similar to `\Google\Site_Kit\Tests\Modules\Analytics_4Test::test_handle_provisioning_callback__gteSupport` but with a failing call to `tagmanager/v2/accounts/containers:lookup` to ensure no exception is unhandled

## QA Brief

This issue primarily fixes the issue observed in the GA account provisioning flow when the `gteSupport` feature is enabled and the GTM scope has not been granted.

One of the ways it addresses this is by requesting this scope as part of the OAuth flow that starts the account creation process.

The key detail which made this error possible to encounter was connecting the Analytics module as part of the initial setup which will not include the GTM scope in the initial OAuth flow because it is conditional on the `gteSupport` feature flag which will not be known at that point (the site isn't connected to the service yet).

A secondary issue was observed and fixed as part of the changes here which could result in a non-blocking error showing up in the GA setup flow because Google tag settings were being requested when they shouldn't have been.

Note: feature flags are not erased from the site as part of a normal reset, so if reusing a site for testing, you'll need to do a persistent reset or delete the `googlesitekitpersistent_remote_features` option in the database manually.

This can be done by adding the following filter which will make the normal reset mechanism include persistent fields.
```
add_filter( 'googlesitekit_hard_reset_enabled', '__return_true' );
``` 

## Changelog entry

* Fix a potential error when creating a new Analytics account.

**Comment by adamdunnage** [comment]
@mohitwp I was just about to open an issue on this myself as I first encountered this today and we have since had a [single report ](https://wordpress.org/support/topic/error-aware-today-when-i-finished-install-site-kit/#post-16788835)in the support forum of this. I am seeing this on the latest available version 1.101.0.

A critical error message is being displayed on completion of Site Kit setup. This is happening when completing setup and clicking on the **Go to my Dashboard** button. When testing, This looks to be happening on almost every new setup with a single exception while testing.  This is reproducible on 1.101.0 but not on the previous 1.99.0. 

After receiving the critical error on screen, you can refresh the page which will take you back to the Site Kit dashboard and prompt with the grant extra permissions banner for GTE support. When granting the permissions Site Kit appears to work as expected with no further issues.

This seems to be an issue introduced yesterday/today and possibly related to a recent service deployment.

Many different console errors displayed throughout the setup process which can be seen in the captured recording. The specific error after completing the setup is as follows:

`[An error of type E_ERROR was caused in line 100 of the file /home/customer/www/dplustire.com/public_html/wp-content/plugins/google-site-kit/third-party/google/apiclient/src/Http/REST.php. Error message: Uncaught Google\Site Kit Dependencies\Google\Service\Exception: { “error”: { “code”: 403, “message”: “Request had insufficient authentication scopes.”, “errors”: [ { “message”: “Insufficient Permission”, “domain”: “global”, “reason”: “insufficientPermissions” } ], “status”: “PERMISSION_DENIED”, “details”: [ { “@type”: “type.googleapis.com/google.rpc.ErrorInfo”, “reason”: “ACCESS_TOKEN_SCOPE_INSUFFICIENT”, “domain”: “googleapis.com”, “metadata”: { “service”: “tagmanager.googleapis.com”, “method”: “container tag.apiary_v2.TagManagerServiceV2.LookupContainer”](https://www.google.com/search?newwindow=1&sxsrf=APwXEdcVGFEiJXDKwuBYOmla9ztB0HOfHA:1685696973442&q=An+error+of+type+E_ERROR+was+caused+in+line+100+of+the+file+/home/customer/www/dplustire.com/public_html/wp-content/plugins/google-site-kit/third-party/google/apiclient/src/Http/REST.php.+Error+message:+Uncaught+Google%5CSite+Kit+Dependencies%5CGoogle%5CService%5CException:+%7B+%22error%22:+%7B+%22code%22:+403,+%22message%22:+%22Request+had+insufficient+authentication+scopes.%22,+%22errors%22:+%5B+%7B+%22message%22:+%22Insufficient+Permission%22,+%22domain%22:+%22global%22,+%22reason%22:+%22insufficientPermissions%22+%7D+%5D,+%22status%22:+%22PERMISSION_DENIED%22,+%22details%22:+%5B+%7B+%22@type%22:+%22type.googleapis.com/google.rpc.ErrorInfo%22,+%22reason%22:+%22ACCESS_TOKEN_SCOPE_INSUFFICIENT%22,+%22domain%22:+%22googleapis.com%22,+%22metadata%22:+%7B+%22service%22:+%22tagmanager.googleapis.com%22,+%22method%22:+%22container+tag.apiary_v2.TagManagerServiceV2.LookupContainer%22&spell=1&sa=X&ved=2ahUKEwi50rPKnqT_AhUhLkQIHfWzBmIQBSgAegQIEhAB)`

Here is a recording of the experience:

https://www.loom.com/share/b200cace1787468eb088cfc5ccdcd2f3

**Comment by wpdarren** [comment]
I receive a slightly different fatal error than @mohitwp but yes, this appears when you attempt to come back to the dashboard from setting up an Analytics account.  This occurs on the latest and main branches, so not related to the plugin but the service side, I suspect. 

![image](https://github.com/google/site-kit-wp/assets/73545194/db1aafe6-2ac8-4409-a8cb-dc00e6384709)

**Comment by tofumatt** [comment]
I'm assuming this is related, but if I set up a new site with existing properties in Analytics (both an existing GA4 property and existing UA property), I get this message:

<img width="1408" alt="CleanShot 2023-06-02 at 21 21 57" src="https://github.com/google/site-kit-wp/assets/90871/f893483c-df6d-49eb-9a3f-7646145b4f4e">

It's not a fatal error but I suspect related. If I click the button, the message goes away.

**Comment by aaemnnosttv** [comment]
@tofumatt this is expected when `gteSupport` is enabled for a site that already has GA set up. You shouldn't encounter it after setting up GA with `gteSupport` enabled to start with though. It is expected that a secondary admin would see this though, similar to how any new required scopes would require oauth/granting when another admin activates a module that requires scopes.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

This is an interesting one to try and explain!  

I was running through this on a new site (never had Site Kit set up before) with a Google account that had previously had Site Kit access granted, but with a new Google Analytics account. When I was redirected to the dashboard no success notification banner appeared and the CTA for Analytics was 'Complete setup'.  When I went to Settings, Analytics was connected successfully. I then went back to the dashboard and everything looked fine. 

**Additional information from further testing:**

All of these scenarios were tested on new sites each time - on main branch. 

1. When I ran through the issue steps above, but with `GA4Reporting` feature flag enabled, no Analytics success notification banner appeared but Analytics did appear on the dashboard, so the 'Complete setup' CTA did not appear. To clarify, this is when using a Google Analytics that had previously had Site Kit access granted. 

2.  When I used a Google account that had Site Kit removed from within the Security tab in Google account settings, this issue did not occur.  Everything was set up as expected. I went through all of the oAuth steps and gave permission. This is with `GA4Reporting` and `GTE Support` feature flags enabled and also `GA4Reporting disabled`. 

Here's a video of the experience. I'd recommend jumping to around 1m 13s to see the behaviour. 

https://github.com/google/site-kit-wp/assets/73545194/382de89c-bfcc-46fc-b123-d02c2c5a0448

I would be interested to know if anyone can re-create this.

**Comment by mohitwp** [comment]
@wpdarren 

Tested this on main environment using tester plugin version 1.9.4.

I tested this 3 times on new instaWP site with new analytics a/c and **not able to recreate** the issue which you are getting.

**Verified cases-** 
1) With 'ga4Reporting' feature flag disabled. 
2) With 'ga4Reporting' feature flag enabled and completed Site kit setup in one flow i.e., connected both search console and analytics during initial set up.
3) With 'ga4Reporting' feature flag Disabled. Connected search console first during initial set up and later connected analytics separately.

**Comment by aaemnnosttv** [comment]
> When I was redirected to the dashboard no success notification banner appeared and the CTA for Analytics was 'Complete setup'. When I went to Settings, Analytics was connected successfully. I then went back to the dashboard and everything looked fine.

That's strange. I just tested this out and got the success banner after creating a new GA account as part of the initial setup. 
<img width="1760" alt="image" src="https://github.com/google/site-kit-wp/assets/1621608/339ef92f-baf5-4ede-94d1-2cc03e4d29d8">

> To clarify, this is when using a Google Analytics that had previously had Site Kit access granted.

This shouldn't matter unless you're describing a scenario where you're reusing the same site for a test, where there could be residual client side storage between runs that could prevent the banner from appearing.

> When I used a Google account that had Site Kit removed from within the Security tab in Google account settings, this issue did not occur.

This shouldn't really be a factor for us but I suppose anything is possible;

Can you give this a test on a different host perhaps?

The only reason you should not see the success banner after creating a new GA account is if there was some error in the process, in which case you'd still end up on the dashboard, but with an `error_code` in the URL. That sounds reasonable given the first account where you saw GA in what sounded like an unconnected state although for it to then become connected after refreshing doesn't make any sense.

---

For testing this one, it's worth noting that we don't want to override the granted feature flags entirely because when the GTE flag is granted is significant. For testing GA4 reporting, we can use the Tester's option to enable something only if it's not enabled yet (rather than force overriding all) since all new sites would get `gteSupport` granted right away. I should have mentioned that in the QAB. I don't think that this would explain the weirdness @wpdarren reported though.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

I am unsure why I was seeing the behaviour reported above in the screencast. I came back a few hours later and I have been unable to recreate it.  @mohitwp and @adamdunnage  were also unable to recreate it.   Baffled!

**Verified:**

- Tested with a new site with `GA4Reporting` disabled, and I am able to go through the setup flow with Analytics included. When I get to the "Go to Dashboard' button, we no longer see the fatal error.  I tested this on latest release and main, just to make sure it is fixed for users on 1.101.0. 
- Tested with a new site with `GA4Reporting` enabled, and I am able to go through the setup flow with Analytics included. When I get to the "Go to Dashboard' button, we no longer see the fatal error.  I tested this on latest release and main, just to make sure it is fixed for users on 1.101.0. 

- I also tested with the `GTESupport` feature flag disabled at initial set up, and when the set up was complete the feature flag is enabled in the site health information, as per Evan's comment above. 
- During the setup flow for the scenarios above, no console warning or error messages appeared. 

For all of the testing above, I tried this with the Site Kit removed from security settings in the Google account used, this was so that the checkboxes appear during oAuth and I gave permission for each test.   

As a secondary round of testing, I set up a new site without clearing the permissions to ensure no problems occurred here - it's possible that more than one site could be used per Google account.
